[Document Intelligence] Release 1.0.0 GA (#31746)
Azure/azure-sdk-for-js: sdk/documentintelligence/ai-document-intelligence-rest/CHANGELOG.md

> ### Features Added

Reviewer: If the release date is known, should we set it?

> - Exports method `streamToUint8Array` to support converting a `NodeJS.ReadableStream` to a `Uint8Array`. This is necessary to read the pdf and png responses from the results of an analysis.

Reviewer: ```suggestion
- Adds `streamToUint8Array`, a convenience function that buffers a `NodeJS.ReadableStream` in a `Uint8Array`. It can be used to read the pdf and png responses from the results of an analysis.
```

Reviewer: I wonder why can't we stream the response directly to disk instead of buffering it first in memory?

Author: This function is to mainly being served from this package to close the gap of not supporting `Uint8Array` return type though it is being offered.

  ```js
  // Example for the figures api that provides an image output
  const output = await client
    .path(
      "/documentModels/{modelId}/analyzeResults/{resultId}/figures/{figureId}",
      "prebuilt-layout",
      resultId,
      figureId
    )
    .get()
    .asNodeStream();

  if (output.status !== "200" || !output.body) {
    throw new Error("The response was unexpected.");

>   }

Reviewer: should we add an example of how to use the function being advertised here?

  ```

> - New `parseResultIdFromResponse` method has been added to parse `operationId` from the initial response of batch analysis requests which can later be used to get the analysis results.

Reviewer: Fix that it is not a method and rewording for clarification:
```suggestion
- Adds `parseResultIdFromResponse`, a convenience function that extracts the `operationId` from the batch analysis response.
```

  ```js
  // Example
  const initialResponse = await client
    .path("/documentModels/{modelId}:analyzeBatch", "prebuilt-layout")
    .post({
      contentType: "application/json",
      body: {
        azureBlobSource: {
          containerUrl: batchTrainingFilesContainerUrl(),
        },
        resultContainerUrl: batchTrainingFilesResultContainerUrl(),
        resultPrefix: "result",
      },
    });

  if (isUnexpected(initialResponse)) {
    throw initialResponse.body.error;
  }
  const batchResultId = parseResultIdFromResponse(initialResponse);

  const response = await client
    .path(
      "/documentModels/{modelId}/analyzeBatchResults/{resultId}",
      "prebuilt-layout",
      batchResultId
    )
    .get();
  ```

> - **Interface Properties:**

Reviewer: nitpick: clarifies the point:
```suggestion
- Changes the following interfaces as follows:
```

  - `AnalyzeBatchDocumentsBodyParam`:

>     - Changed `body` from optional to required.
>   - `AnalyzeBatchOperationOutput`:
>     - Added `resultId`.
>   - `AnalyzeDocumentBodyParam`:
>     - Changed `body` from optional to required.
>   - `DocumentClassifierDetailsOutput`:
>     - Added `modifiedDateTime`.
>   - `DocumentModelDetailsOutput`:
>     - Added `modifiedDateTime`.

Reviewer: ```suggestion
    - Changes `body` from optional to required.
  - `AnalyzeBatchOperationOutput`:
    - Adds `resultId`.
  - `AnalyzeDocumentBodyParam`:
    - Changes `body` from optional to required.
  - `DocumentClassifierDetailsOutput`:
    - Adds `modifiedDateTime`.
  - `DocumentModelDetailsOutput`:
    - Adds `modifiedDateTime`.
```

- **New Interfaces:**
  Define query parameters for various document analysis requests, offering flexibility with options like `style` and `explode`:
  - **AnalyzeBatchDocumentsFeaturesQueryParam**: Customizable `style` and `explode` options for batch document analysis, with a value of type `DocumentAnalysisFeature[]`.
  - **AnalyzeBatchDocumentsOutputQueryParam**: Customizable `style` and `explode` options for batch document analysis, with a value of type `AnalyzeOutputOption[]`.
  - **AnalyzeBatchDocumentsQueryFieldsQueryParam**: Customizable `style` and `explode` options for batch document analysis, with a value of type `string[]`.
  - **AnalyzeDocumentFeaturesQueryParam**: Customizable `style` and `explode` options for single document analysis, with a value of type `DocumentAnalysisFeature[]`.

>   - **AnalyzeDocumentFromStreamFeaturesQueryParam**: Customizable `style` and `explode` options for document analysis from stream, with a value of type `DocumentAnalysisFeature[]`.

Reviewer: ```suggestion
- Introduces new interfaces to define query parameters for document analysis requests, allowing customizable `style` and `explode` options:
  - **AnalyzeBatchDocumentsFeaturesQueryParam**: Accepts DocumentAnalysisFeature[] values.
  - **AnalyzeBatchDocumentsOutputQueryParam**: Accepts AnalyzeOutputOption[] values.
  - **AnalyzeBatchDocumentsQueryFieldsQueryParam**: Accepts string[] values.
  - **AnalyzeDocumentFeaturesQueryParam**: Accepts DocumentAnalysisFeature[] values.
  - **AnalyzeDocumentFromStreamFeaturesQueryParam**: Accepts DocumentAnalysisFeature[] values.

### Breaking Changes

> - Removes the `poller.getOperationId()` for a given polling operation in favour of the global `parseResultIdFromResponse` method.

Reviewer: Clarifies that it is not a method:
```suggestion
- Removes the `poller.getOperationId()` for a given polling operation. Use `parseResultIdFromResponse` to extract the `operationId` directly.
```

### Other Changes

> The following types were renamed

Reviewer: ```suggestion
The following types are renamed
```

        throw new Error("The response was unexpected.");
      }
  ```
- New `parseOperationIdFromResponse` method has been added to parse `operationId` from the initial response of batch analysis requests which can later be used to get the analysis results.
  ```js
      // Example
      const initialResponse = await client
        .path("/documentModels/{modelId}:analyzeBatch", "prebuilt-layout")
        .post({
          contentType: "application/json",
          body: {
            azureBlobSource: {
              containerUrl: batchTrainingFilesContainerUrl(),
            },
            resultContainerUrl: batchTrainingFilesResultContainerUrl(),
            resultPrefix: "result",
          },
        });

      if (isUnexpected(initialResponse)) {
        throw initialResponse.body.error;
      }

>       const batchOperationId = parseOperationIdFromResponse(initialResponse);

Reviewer 2: Dont know if it impacts JS, but we are exposing the operationId as resultId in Java.

Author: is that a rename?

Reviewer 3: Here's what Paul said during archboard:
> OperationDetails.operationId is for :build, :compose, and :copyTo operations.  There, operationId is more appropriate than resultId.  AnalyzeBatchResultOperation and AnalyzeResultOperation are LRO :analyze.  Here, resultId is more semantically accurate.

I'm not sure if it was named differently before, but the property is defined as `resultId` in the spec today.

Author: Updated the global method to `parseResultIdFromResponse`
